### PR TITLE
Fix type of Customer.sources

### PR DESCRIPTION
**Pull request by LouiseBC**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [clicky](https://stripe.com/docs/api/sources/object)

There is certainly still room for improvement here. For instance, it's entirely possible to specify whether the `receiver` or `redirect` objects are present on a source object. However, I'm constrained in time and I think this is a step in the right direction, at least.

Also please excuse the linting changes. I'm aware that they make the real changes a bit harder to spot, but I think they were good changes.

### CODE REVIEW

File: types/stripe/index.d.ts [hunk 1]
```diff
@@ -5209,6 +5209,258 @@ declare namespace Stripe {
          * card.
          */
         type ISourceCreationOptionsExtended = string | cards.ISourceCreationOptionsExtended;
+
+        type ISource = ISourceBase & (ICardSource | IOtherSource)
+        type ISourceTypes = "ach_credit_transfer" | "ach_debit" | "alipay" | "bancontat" | "card" | "card_present" | "eps" | "giropay" | "ideal" | "multibanco" | "p24" | "paper_check" | "sepa_credit_transfer" | "sepa_debit" | "sofort" | "three_d_secure";
+
+        interface IOtherSource extends ISourceBase {
+            type: ISourceTypes;
+            [k: string]: any;
```
Comment (squirly) [inline types/stripe/index.d.ts]: This is too broad. Only `ISourceType` should be allowed as keys here. And the value's type is always object like.
```ts
[source_data: ISourceType]: Record<string, any>
```
This key may need to be optional.

File: types/stripe/index.d.ts [hunk 2]
```diff
@@ -5209,6 +5209,258 @@ declare namespace Stripe {
          * card.
          */
         type ISourceCreationOptionsExtended = string | cards.ISourceCreationOptionsExtended;
+
+        type ISource = ISourceBase & (ICardSource | IOtherSource)
+        type ISourceTypes = "ach_credit_transfer" | "ach_debit" | "alipay" | "bancontat" | "card" | "card_present" | "eps" | "giropay" | "ideal" | "multibanco" | "p24" | "paper_check" | "sepa_credit_transfer" | "sepa_debit" | "sofort" | "three_d_secure";
```
Comment (squirly) [inline types/stripe/index.d.ts]: This name should be singular. Usually only array or set types are plural.

**Comment by LouiseBC** [comment]
Thanks for the pointers. I've fixed it as much as I can. I wanted to make it so that the compiler would understand that if type === "alipay" that the key "alipay" would be available, but it doesn't seem possible.

**Comment by squirly** [comment]
@LouiseBC Ican think of two ways to do this. The result is very similar. On is to use a Mapped Type to enumerate the possibilities then use the type accessor `[]` to Union them together.
```ts
// @ Todo make example when at a computer
```
The other option is to manually type each interface and merge them as a type Union as you have already started to do.
```ts
type ISource = ICardSource | INameHereSource | ... ;
type SourceType = ISource['type'];

Interface INameHereSource extends IBaseSource {
  type: 'name_here';
  name_here: {
    [key: string]: any;
  }
}
```
I feel the latter is more desirable, because it will allow for easy addition of the actual keys in future pull requests.

**Comment by armanio123** [comment]
@LouiseBC @squirly Is there anything else to address before merging or it's good as is?

**Comment by LouiseBC** [comment]
@squirly Sorry for my delayed response! And thank you for your suggestions. The latter is indeed preferable, but afaik not all source type information is documented, and I don't have the time to add them all now even if it was. The former suggestion does not work as-is:
```typescript
        type ISource = ICardSource | {
             // error on [Type]
             // - A computed property name in a type literal must refer to an expression whose type is a literal type or a 'unique symbol' type.
             // -'Type' only refers to a type, but is being used as a value here.
            [Type in IOtherSourceType]: {type: Type, [Type]: Record<string, any>} & ISourceBase
          }[IOtherSourceType]
```

I've tried a lot of variations to accomplish the desired behaviour but neither I or my colleague managed to get it to work.
As an aside, could you explain the type accessor syntax? I've never seen it used in that way

**Review by squirly — COMMENTED** [review]
I believe this can be achieved with
```ts
type OtherSourceType = "ach_credit_transfer" | "ach_debit" | "alipay" | "bancontat" | "card_present" | "eps" | "giropay" | "ideal" | "multibanco" | "p24" | "paper_check" | "sepa_credit_transfer" | "sepa_debit" | "sofort" | "three_d_secure";
type IGenericSource<SourceType extends string> = ISourceBase & Record<SourceType, string> & {
    type: SourceType;
}

type ISource = ICardSource |
    { [Type in OtherSourceType]: IGenericSource<Type> }[OtherSourceType];
```

[Working Example](https://www.typescriptlang.org/play/#src=interface%20ISourceBase%20%7B%0D%0A%20%20%20%20name%3A%20'test'%3B%0D%0A%20%20%20%20%2F%2F%20Trimmed%20for%20simplicity%0D%0A%7D%0D%0A%0D%0Ainterface%20ICardSource%20extends%20ISourceBase%20%7B%0D%0A%20%20%20%20type%3A%20'card'%3B%0D%0A%20%20%20%20card%3A%20%7B%0D%0A%20%20%20%20%20%20%20%20last4%3A%20'1234'%3B%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20Trimmed%20for%20simplicity%0D%0A%20%20%20%20%7D%3B%0D%0A%7D%3B%0D%0A%0D%0Atype%20OtherSourceType%20%3D%20%22ach_credit_transfer%22%20%7C%20%22ach_debit%22%20%7C%20%22alipay%22%20%7C%20%22bancontat%22%20%7C%20%22card_present%22%20%7C%20%22eps%22%20%7C%20%22giropay%22%20%7C%20%22ideal%22%20%7C%20%22multibanco%22%20%7C%20%22p24%22%20%7C%20%22paper_check%22%20%7C%20%22sepa_credit_transfer%22%20%7C%20%22sepa_debit%22%20%7C%20%22sofort%22%20%7C%20%22three_d_secure%22%3B%0D%0Atype%20IGenericSource%3CSourceType%20extends%20string%3E%20%3D%20ISourceBase%20%26%20Record%3CSourceType%2C%20string%3E%20%26%20%7B%0D%0A%20%20%20%20type%3A%20SourceType%3B%0D%0A%7D%0D%0A%0D%0Atype%20ISource%20%3D%20ICardSource%20%7C%0D%0A%20%20%20%20%7B%20%5BType%20in%20OtherSourceType%5D%3A%20IGenericSource%3CType%3E%20%7D%5BOtherSourceType%5D%3B%0D%0A%0D%0A()%20%3D%3E%20%7B%0D%0A%20%20%20%20let%20source%3A%20ISource%3B%0D%0A%20%20%20%20switch%20(source.type)%20%7B%0D%0A%20%20%20%20%20%20%20%20case%20'card'%3A%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20return%20source%3B%20%2F%2F%20narrowed%20to%20ICardSource%0D%0A%20%20%20%20%20%20%20%20case%20'ach_credit_transfer'%3A%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20return%20source%3B%20%2F%2F%20narrowed%20to%20IGenericSource%3C%22ach_credit_transfer%22%3E%0D%0A%20%20%20%20%20%20%20%20case%20'ach_debit'%3A%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20return%20source%3B%20%2F%2F%20narrowed%20to%20IGenericSource%3C%22ach_credit_transfer%22%3E%0D%0A%20%20%20%20%20%20%20%20default%3A%0D%0A%20%20%20%20%20%20%20%20%20%20%20%20return%20source%3B%20%2F%2F%20narrowed%20to%20IGenericSource%3C%22alipay%22%3E%20%7C%20IGenericSource%3C%22bancontat%22%3E%20%7C%20IGenericSource%3C%22card_present%22%3E%20%7C%20IGenericSource%3C%22eps%22%3E%20%7C%20IGenericSource%3C%22giropay%22%3E%20%7C%20IGenericSource%3C%22ideal%22%3E%20%7C%20...%206%20more%20...%20%7C%20IGenericSource%3C...%3E%0D%0A%20%20%20%20%7D%0D%0A%7D%3B%0D%0A)

File: types/stripe/index.d.ts [hunk 3]
```diff
@@ -5209,6 +5209,256 @@ declare namespace Stripe {
          * card.
          */
         type ISourceCreationOptionsExtended = string | cards.ISourceCreationOptionsExtended;
+
+        type IOtherSourceType = "ach_credit_transfer" | "ach_debit" | "alipay" | "bancontat" | "card_present" | "eps" | "giropay" | "ideal" | "multibanco" | "p24" | "paper_check" | "sepa_credit_transfer" | "sepa_debit" | "sofort" | "three_d_secure";
```
Comment (squirly) [inline types/stripe/index.d.ts]: Remove the `I`, it is for interface types.

**Comment by AviVahl** [comment]
fixed in https://github.com/DefinitelyTyped/DefinitelyTyped/pull/35633

**Comment by orta** [comment]
#35633 was merged

**Pull request closed without merging**